app/services/DriveService.scala

```diff
@@ -34,34 +35,90 @@ object DriveService {
       .setTeamDriveId(Config.teamDriveId)
 
     request.execute().getFiles.asScala.toList
-      .map(file => { // extract this into a pure function we can test
-
-        // raw model file
-        // function transforms
-        // output target models
-
-        // The inline Try will handle nulls returned from Java library
-        val customProps = Try {
-          file.getProperties.asScala.toMap
-        }.toOption.getOrElse(Map.empty)
-        val thumbnailLink = Try {
-          file.getThumbnailLink
-        }.toOption.getOrElse("")
-        val exportLinks = Try {
-          file.getExportLinks.asScala.toMap
-        }.toOption.getOrElse(Map.empty)
-
-        File(
-          id = file.getId,
-          title = file.getName,
-          output = file.getMimeType,
-          outputPreview = file.getWebViewLink, // TODO - returning editable links - need to process to preview
-          outputDownload = file.getWebContentLink, // TODO - returning nulls
-          outputThumbnail = thumbnailLink, // TODO - returning 404s
-          exportLinks = exportLinks,
-          customProperties = customProps
-        )
-      })
+      .map(file =>
+        // transformToResearchRecord(extractRawFile(file)) // define functions with def
+        (extractRawFile andThen transformToResearchRecord)(file) // define functions with val
+    )
+  }
+
+  val extractRawFile = (fileStream: File) => {
```

Comment:
Hi. My understanding was that AndThen only exists on `val` functions not `def` functions, and we can't compile when we use def here.  Perhaps we are doing it wrong? 🤔

Reply:
Oh gosh, I missed that - good point. This is an example of Scala being annoying. With `def` functions you can `andThen` them if you follow them with an underscore:

```scala
def addOne(i: Int): Int = i + 1
def addTwo(i: Int): Int = i + 2

addOne andThen addTwo  // error
addOne _ andThen addTwo _
// in fact, only the first function needs the underscore
addOne _ andThen addTwo
```

But obviously this is awful.

It's a trade-off between:

* write `val` functions and use `andThen`
* write `def` functions and compose by hand

Remember, these two are exactly equal:

```scala
val fn1 = (arg: Type) => body
val fn2 = (arg: Type) => Body
(fn1 andThen fn2)(x)

def fn1(arg: Type) = body
def fn2(arg: Type) = body
fn2(fn1(x))
```
The `def` approach is more conventional for the function definition, and more compact. It's also much easier to include the return type. Adding return types to the val would mean writing the full signature out again, which is verbose and duplicate-y. Compare the following:

```scala
val fn1: Type => Type = (arg: Type) => body
def fn1(arg: Type): Type = body
```
As for the composition bit, I'd probably favour `andThen` if we're creating a new function that is the composition of two others without actually giving it a value:
```scala
val add3 = add1 andThen add2
```

But when you're directly applying it (as in the example above), it's just as easy to be explicit and doing so is much shorter as well `f(g(x))`.

This is a matter of taste! More detail than is required for this very minor point, sorry. It's a tricky edge case that comes from Scala being a hybrid of functional and OO. There are times like this where its functional chops (e.g. having nice syntax for function composition) are a bit hamstrung.

Reply:
you can do (extractRawFile _) to turn it into a value (of Function1) rather than a method (def).  But you're reaching into scala's limitations of syntax when it comes to doing more FP!  I will let you decide which is clearer.